lib/rspec/mocks/error_generator.rb

```diff
@@ -268,6 +268,18 @@ def unexpected_arguments_message(expected_args_string, actual_args_string)
       def error_message(expectation, args_for_multiple_calls)
         expected_args = format_args(expectation.expected_args)
         actual_args = format_received_args(args_for_multiple_calls)
+
+        if RSpec::Support::RubyFeatures.kw_args_supported? && expected_args == actual_args
+          expected_hash = expectation.expected_args.last
+          actual_hash = args_for_multiple_calls.last.last
+          if Hash === expected_hash && Hash === actual_hash
+            if Hash.ruby2_keywords_hash?(expected_hash) ^ Hash.ruby2_keywords_hash?(actual_hash)
+              actual_args += Hash.ruby2_keywords_hash?(actual_hash) ? " (keyword arguments)" : " (options hash)"
```

Comment:
Note: `Hash.ruby2_keywords_hash?` will only recognize hashes from a  `ruby2_keywords` method/block, or a hash explcitly marked with `Hash.ruby2_keywords_hash`, but e.g., not `kw` in `def m(*args, **kw); kw; end`.
I guess `ruby2_keywords` is used somewhere before and so it just works here (is that in RSpec code?), but it's a bit subtle.

If `rspec-mocks` would store positional and kwargs arguments in separate fields (only on Ruby 3+, before can't separate them reliably) then I think it would be nicer and more reliable to check.

Reply:
@eregon what you are saying is right, but I'm not trying to display whether a hash was keyword args 100% of the times, I only want to display it when two hashes are identical except for the `ruby2_keywords_hash?`.

#1394 add this code: 

```ruby
        if RUBY_VERSION >= "3"
          # if both arguments end with Hashes, and if one is a keyword hash and the other is not, they don't match
          if Hash === expected_args.last && Hash === actual_args.last
            if !Hash.ruby2_keywords_hash?(actual_args.last) && Hash.ruby2_keywords_hash?(expected_args.last)
              return false
            end
          end
        end
```

And it's this case I want to display. So yeah, my PR only apply to `3+` like the previous PR.

Reply:
`Hash.ruby2_keywords_hash?` still can't tell if a Hash was used as keyword arguments, except for the case we are sure the keyword arguments were passed to a `ruby2_keywords` method, and that Hash then arrives it, not through another call (which would remove the flag IIRC).
Is that guaranteed here that `expected_hash` and `actual_hash` come from `ruby2_keywords` methods?

In other words, I think the current patch might not be able to show the helpful kwargs/non-kwargs suffix for things like:
```ruby
def foo(*args, **kwargs)
  ...
end

foo({a: 1})
vs
foo(a: 1)
```
but given this is rspec-mocks maybe it could because it might always control the definition of `foo` and always use `ruby2_keywords` there?

Reply:
> Is that guaranteed here that expected_hash and actual_hash come from ruby2_keywords methods?

if `expected_hash == actual_hash` then AFAIK yes that's the only way. Otherwise we wouldn't be printing an error.

Of course there might be some weird edge cases if you monkey patch `Hash#==` but...

Reply:
I see, there is similar logic in https://github.com/mame/rspec-mocks/blob/865ea3a795e0e56309b823e76e3f65eff5cfa448/lib/rspec/mocks/argument_list_matcher.rb#L63-L70
I think a `# See #args_match?` comment would be helpful.

Also the check doesn't necessarily seem correct, it's only in one direction, because passing kwargs to a method not taking kwargs is fine: https://github.com/mame/rspec-mocks/blob/865ea3a795e0e56309b823e76e3f65eff5cfa448/lib/rspec/mocks/argument_list_matcher.rb#L66
(the comment is unfortunately outdated/not matching the code)

I'd suggest to extract the check to a separate method, then it's clear it's the same condition.